### DISCUSSION TITLE
Set time of the initial condition (IC)

### DISCUSSION THREAD

**Discussion by igoraroliveira** [lululxvi/deepxde#1093] [Q&A]
Hello,
I'm having trouble setting a specific time for my IC, that time should be in the middle of my time domain.
I don't seem to find an example that does something like that.

I'm looking forward to active something similar to this image: (time domain in the x-axis, space domain in the y-axis )

![image](https://user-images.githubusercontent.com/110116122/208793234-3672b45e-d826-449d-b557-2cb4dd83f118.png)

For example, I tried something like:

```
def init_cond_u(x):
    return np.exp(x[:, 0:1])

def time_boundary(x, on_initial):
    return on_initial and np.isclose(x[1], 0)

ic = dde.icbc.IC(geomtime, init_cond_u, lambda _, time_boundary, component=0)
```

But it doesn't seems to work, since it is obviously not how I should do that.
Thanks in advance!

**Comment by lululxvi** [comment]
```
def time_boundary(x, on_initial):
    return np.isclose(x[1], 0)

ic = dde.icbc.IC(geomtime, init_cond_u, time_boundary, component=0)
```

**Reply by igoraroliveira** [reply]
> ```
> def time_boundary(x, on_initial):
>     return np.isclose(x[1], 0)
> 
> ic = dde.icbc.IC(geomtime, init_cond_u, time_boundary, component=0)
> ```
Thanks for the reply, and the amazing work in the deepxde!

I tried the suggested solution and, to make it work I also needed to specify a "atol" (The absolute tolerance parameter) in the np.isclose( ) argument. Since the default "atol" failed to find data points with x[1] coordinate close enough to 0.
What worked for me is as follows:

```
def time_boundary(x, on_initial):      
    return np.isclose(x[1], 0.0, atol=5e-03)

ic_u = dde.icbc.IC(geomtime, init_cond_u, time_boundary, component=0)
```
With some test I found that 
`return np.isclose(x[1], 0.0, atol=5e-03)`
returned about 100 times "True",  meaning that enough initial condition points were generated. 

My data was specified as follows: 

```
data = dde.data.TimePDE(
    geomtime,
    pde,
    [ic_u],
    num_domain=10000,
    num_boundary=0,
    num_initial=100,
    train_distribution="pseudo",
)
```
Where the geomtime was defined as:

```
x_lower = -5
x_upper = 5
t_lower = -0.5
t_upper = 0.5

space_domain = dde.geometry.Interval(x_lower, x_upper)
time_domain = dde.geometry.TimeDomain(t_lower, t_upper)
geomtime = dde.geometry.GeometryXTime(space_domain, time_domain)
```
My question is: is this the best way to do that? 

I feel like I'm adding unnecessary error to my initial condition, since it is being informed to my neural network with an initial error, something like that:

```math
y(x,t_0 \pm \delta \sigma) = IC,  \quad \quad t_0 = 0,  \quad \quad \sigma= 5e-3
```
Where $t_0$ represents the time of my initial condition, and some error in the order of 5e-3 is present around the initial condition time.

If this is the right way to define my initial condition, how should I specify an appropriate tolerance to my np.isclose( ) argument?

Thanks for the help!

**Reply by lululxvi** [reply]
It is good. No problem at all.
